### ISSUE TITLE
同一个id的相似度有没有办法提高呢

### ISSUE THREAD

**Issue by Phxmark** [deepinsight/insightface#825]
我用faces_glintasia,arcface训练了一个模型，同一个id的相似度在0.7左右，请问有什么办法可以提高这个相似度吗

**Comment by MagnetStone** [comment]
相似度计算公式贴出来吧，便于分析

**Comment by Phxmark** [comment]
> 相似度计算公式贴出来吧，便于分析
用的与源码中一致的代码
def get_feature(self, aligned):
    input_blob = np.expand_dims(aligned, axis=0)
    data = mx.nd.array(input_blob)
    db = mx.io.DataBatch(data=(data,))
    self.model.forward(db, is_train=False)
    embedding = self.model.get_outputs()[0].asnumpy()
    embedding = sklearn.preprocessing.normalize(embedding).flatten()
    return embedding
用上述函数得到特征f1,f2。
然后计算相似度sim = np.dot(f1, f2.T)

**Comment by haoa060101** [comment]
嗯，我也遇到了这个问题！
同一个ID人脸姿势变化稍大，计算欧式距离，相识度很难提升到0.9以上！

**Comment by mohuazheliu** [comment]
try to calculate the angle between the two feature

**Comment by bleakie** [comment]
> try to calculate the angle between the two feature

What do you mean?

**Comment by mohuazheliu** [comment]
Not Euclidean distance, please to calculate cosine distance

**Comment by bleakie** [comment]
> Not Euclidean distance, please to calculate cosine distance

we used is np.dot(f1, f2.T)，That's the cosine distance

**Comment by mohuazheliu** [comment]
sorry ,I know a little about python,Maybe the result of dot need to be  divided by norm of two feature

**Comment by FaceGg** [comment]
#你试试这么算算，看看行不行
import numpy as np

def cos_sim(f1, f2):
    vector_a = np.mat(f1)
    vector_b = np.mat(f2)
    num = float(f1 * f2.T)
    denom = np.linalg.norm(f1) * np.linalg.norm(f2)
    cos = num / denom
    sim = 0.5 + 0.5 * cos
    return sim

**Comment by FaceGg** [comment]
#你试试这么算算，看看行不行
import numpy as np
def cos_sim(f1, f2):
f1 = np.mat(f1)
f2 = np.mat(f2)
num = float(f1 * f2.T)
denom = np.linalg.norm(f1) * np.linalg.norm(f2)
cos = num / denom
sim = 0.5 + 0.5 * cos
return sim

**Comment by FaceGg** [comment]
> 我用faces_glintasia,arcface训练了一个模型，同一个id的相似度在0.7左右，请问有什么办法可以提高这个相似度吗
#你试试这么算算，看看行不行
import numpy as np
def cos_sim(f1, f2):
f1 = np.mat(f1)
f2 = np.mat(f2)
num = float(f1 * f2.T)
denom = np.linalg.norm(f1) * np.linalg.norm(f2)
cos = num / denom
sim = 0.5 + 0.5 * cos
return sim

**Comment by FaceGg** [comment]
> 我用faces_glintasia,arcface训练了一个模型，同一个id的相似度在0.7左右，请问有什么办法可以提高这个相似度吗

#你试试这么算算，看看行不行
import numpy as np
def cos_sim(f1, f2):
f1 = np.mat(f1)
f2 = np.mat(f2)
num = float(f1 * f2.T)
denom = np.linalg.norm(f1) * np.linalg.norm(f2)
cos = num / denom
sim = 0.5 + 0.5 * cos
return sim

**Comment by liutianyuan** [comment]
> 我用faces_glintasia,arcface训练了一个模型，同一个id的相似度在0.7左右，请问有什么办法可以提高这个相似度吗

请问您问题解决了吗？  @Phxmark

**Comment by skybocai** [comment]
代码里给出的余弦距离取值范围是-1《 sim《1， 所以需要给sim 进行归一化 sim = 0.5+ 0.5*sim 之后得到的余弦相似度0<sim<1这时候相似度应该是提高了的

**Comment by paleomoon** [comment]
> 

But in this way, different person usually get similarity larger than 0.5. Is there a better way to mapping cosine to [0,1]?
